File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/drops/impl/EcoFastCollatedDropQueue.java
package com.willfp.eco.internal.drops.impl;

import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.Accessors;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class EcoFastCollatedDropQueue extends EcoDropQueue {
    public static final Map<Player, CollatedDrops> COLLATED_MAP = new ConcurrentHashMap<>();

    public EcoFastCollatedDropQueue(@NotNull final Player player) {
        super(player);
    }

    @Override
    public void push() {
        CollatedDrops fetched = COLLATED_MAP.get(getPlayer());
        CollatedDrops collatedDrops = fetched == null ? new CollatedDrops(getItems(), getLoc(), getXp()) : fetched.addDrops(getItems()).setLocation(getLoc()).addXp(getXp());
        COLLATED_MAP.put(this.getPlayer(), collatedDrops);
    }

    @ToString
    public static final class CollatedDrops {
        @Getter
        private final List<ItemStack> drops;

        @Getter
        @Setter
        @Accessors(chain = true)
        private Location location;

        @Getter
        private int xp;

        private CollatedDrops(@NotNull final List<ItemStack> drops,
                              @NotNull final Location location,
                              final int xp) {
            this.drops = drops;
            this.location = location;
            this.xp = xp;
        }

        public CollatedDrops addDrops(@NotNull final List<ItemStack> toAdd) {
            drops.addAll(toAdd);
            return this;
        }

        public CollatedDrops addXp(final int xp) {
            this.xp += xp;
            return this;
        }
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/drops/EcoFastCollatedDropQueue.kt
package com.willfp.eco.internal.drops

import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.util.concurrent.ConcurrentHashMap

class EcoFastCollatedDropQueue(player: Player) : EcoDropQueue(player) {
    override fun push() {
        val fetched = COLLATED_MAP[player]

        if (fetched == null) {
            COLLATED_MAP[player] = CollatedDrops(items, location, xp, hasTelekinesis)
        } else {
            fetched.addDrops(items)
            fetched.location = location
            fetched.addXp(xp)
            if (this.hasTelekinesis) {
                fetched.forceTelekinesis()
            }

            COLLATED_MAP[player] = fetched
        }
    }

    class CollatedDrops(
        val drops: MutableList<ItemStack>,
        var location: Location,
        var xp: Int,
        var telekinetic: Boolean
    ) {
        fun addDrops(toAdd: List<ItemStack>): CollatedDrops {
            drops.addAll(toAdd)
            return this
        }

        fun addXp(xp: Int): CollatedDrops {
            this.xp += xp
            return this
        }

        fun forceTelekinesis() {
            telekinetic = true
        }
    }

    companion object {
        val COLLATED_MAP: MutableMap<Player, CollatedDrops> = ConcurrentHashMap()
    }
}